File: pdftransl/rag/embeddings.py
"""Подключаемые эмбеддеры для памяти переводов.

Три уровня, чтобы RAG работал всегда: HashingEmbedder (без
зависимостей, офлайн), SentenceTransformerEmbedder (локальные
нейро-эмбеддинги), ApiEmbedder (любой OpenAI-совместимый
/embeddings).
"""

from __future__ import annotations

import hashlib
import math
import re
import time
from abc import ABC, abstractmethod
from typing import Optional

from pdftransl.config import PipelineConfig


class BaseEmbedder(ABC):
    name: str = "base"
    dim: int = 0

    @abstractmethod
    def embed(self, texts: list[str]) -> list[list[float]]: ...


_TOKEN_RE = re.compile(r"[^\W_]+", re.UNICODE)


class HashingEmbedder(BaseEmbedder):
    """Character 3-gram hashing into a fixed-size normalized vector."""

    name = "hashing"

    def __init__(self, dim: int = 512):
        self.dim = dim

    def _vector(self, text: str) -> list[float]:
        vec = [0.0] * self.dim
        tokens = _TOKEN_RE.findall(text.lower())
        for token in tokens:
            padded = f"^{token}$"
            for i in range(len(padded) - 2):
                gram = padded[i:i + 3]
                digest = hashlib.md5(gram.encode("utf-8")).digest()
                idx = int.from_bytes(digest[:4], "little") % self.dim
                sign = 1.0 if digest[4] % 2 else -1.0
                vec[idx] += sign
        norm = math.sqrt(sum(v * v for v in vec)) or 1.0
        return [v / norm for v in vec]

    def embed(self, texts: list[str]) -> list[list[float]]:
        return [self._vector(t) for t in texts]


class SentenceTransformerEmbedder(BaseEmbedder):
    name = "sentence-transformers"

    def __init__(self, model_name: str = "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"):
        from sentence_transformers import SentenceTransformer  # lazy import

        self.model = SentenceTransformer(model_name)
        self.dim = self.model.get_sentence_embedding_dimension()

    def embed(self, texts: list[str]) -> list[list[float]]:
        vectors = self.model.encode(texts, normalize_embeddings=True)
        return [v.tolist() for v in vectors]


class ApiEmbedder(BaseEmbedder):
    name = "api"

    def __init__(self, base_url: str, model: str, api_key: Optional[str] = None):
        import requests

        self._requests = requests
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.api_key = api_key
        self.dim = 0  # discovered on first call

    def embed(self, texts: list[str]) -> list[list[float]]:
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        last_error = None
        for attempt in range(3):
            try:
                resp = self._requests.post(
                    f"{self.base_url}/embeddings",
                    json={"model": self.model, "input": texts},
                    headers=headers,
                    timeout=(15, 120),
                )
                if resp.status_code in (408, 429, 500, 502, 503, 504):
                    last_error = f"HTTP {resp.status_code}"
                    time.sleep(2 ** attempt)
                    continue
                resp.raise_for_status()
                data = resp.json()["data"]
                # OpenAI-compatible endpoints identify each vector by index;
                # do not silently pair a reordered response with wrong text.
                data = sorted(data, key=lambda item: item.get("index", 0))
                vectors = [item["embedding"] for item in data]
                if len(vectors) != len(texts):
                    raise ValueError("embedding response length mismatch")
                break
            except (self._requests.RequestException, KeyError, TypeError, ValueError) as exc:
                last_error = str(exc)
                if attempt == 2:
                    raise RuntimeError(f"embedding request failed: {last_error}") from exc
                time.sleep(2 ** attempt)
        else:  # pragma: no cover - loop either breaks or raises
            raise RuntimeError(f"embedding request failed: {last_error}")
        if vectors:
            self.dim = len(vectors[0])
        return vectors


def get_embedder(config: PipelineConfig) -> BaseEmbedder:
    """'auto' prefers sentence-transformers if installed, else hashing."""
    import os

    choice = config.embedder
    if choice == "auto":
        try:
            import sentence_transformers  # noqa: F401
            choice = "sentence-transformers"
        except ImportError:
            choice = "hashing"
    if choice == "hashing":
        return HashingEmbedder()
    if choice == "sentence-transformers":
        return SentenceTransformerEmbedder(
            config.embedding_model
            or "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"
        )
    if choice == "api":
        api_key = (
            os.environ.get(config.embedding_api_key_env)
            if config.embedding_api_key_env
            else None
        )
        return ApiEmbedder(
            base_url=config.embedding_base_url or "http://localhost:11434/v1",
            model=config.embedding_model or "nomic-embed-text",
            api_key=api_key,
        )
    raise ValueError(f"Unknown embedder '{choice}'")


def cosine(a: list[float], b: list[float]) -> float:
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a)) or 1.0
    nb = math.sqrt(sum(x * x for x in b)) or 1.0
    return dot / (na * nb)

File: cpp/src/rag/embeddings.cpp
#include "rag/embeddings.h"
#include <QCryptographicHash>
#include <QRegularExpression>
#include <cmath>

namespace pdftransl {

namespace {

const QRegularExpression& tokenRe() {
    // Unicode-буквенно-цифровые последовательности (латиница, кириллица,
    // CJK...); \W/\d исключают пунктуацию и цифры-как-разделители, как в
    // pdftransl.rag.embeddings._TOKEN_RE.
    static const QRegularExpression re(QStringLiteral("[^\\W\\d_]+"),
                                        QRegularExpression::UseUnicodePropertiesOption);
    return re;
}

} // namespace

HashingEmbedder::HashingEmbedder(int dim) : m_dim(dim > 0 ? dim : 128) {}

std::vector<float> HashingEmbedder::embed(const QString& text) const {
    std::vector<float> vec(static_cast<size_t>(m_dim), 0.0f);

    QRegularExpressionMatchIterator tokens = tokenRe().globalMatch(text.toLower());
    while (tokens.hasNext()) {
        const QString token = tokens.next().captured(0);
        const QString padded = QStringLiteral("^") + token + QStringLiteral("$");
        for (int i = 0; i + 3 <= padded.size(); ++i) {
            const QString gram = padded.mid(i, 3);
            const QByteArray digest =
                QCryptographicHash::hash(gram.toUtf8(), QCryptographicHash::Sha256);
            // Первые 4 байта дайджеста -> индекс корзины; следующий байт ->
            // знак вклада (детерминированный псевдослучайный проекционный
            // хеш, как в hashing trick / feature hashing).
            quint32 raw = 0;
            for (int b = 0; b < 4; ++b) {
                raw |= static_cast<quint32>(static_cast<unsigned char>(digest[b])) << (8 * b);
            }
            const int idx = static_cast<int>(raw % static_cast<quint32>(m_dim));
            const float sign = (static_cast<unsigned char>(digest[4]) % 2) ? 1.0f : -1.0f;
            vec[static_cast<size_t>(idx)] += sign;
        }
    }

    double norm = 0.0;
    for (float v : vec) norm += static_cast<double>(v) * v;
    norm = std::sqrt(norm);
    if (norm > 0.0) {
        for (float& v : vec) v = static_cast<float>(v / norm);
    }
    return vec;
}

double cosine(const std::vector<float>& a, const std::vector<float>& b) {
    if (a.empty() || b.empty() || a.size() != b.size()) return 0.0;
    double dot = 0.0, na = 0.0, nb = 0.0;
    for (size_t i = 0; i < a.size(); ++i) {
        dot += static_cast<double>(a[i]) * b[i];
        na += static_cast<double>(a[i]) * a[i];
        nb += static_cast<double>(b[i]) * b[i];
    }
    na = std::sqrt(na);
    nb = std::sqrt(nb);
    if (na <= 0.0 || nb <= 0.0) return 0.0;
    return dot / (na * nb);
}

} // namespace pdftransl
